### PR TITLE
[50_1] New entry for Mogan Draw

### PR DESCRIPTION
**Pull request by Oyyko**

### CODE REVIEW

**Comment by Oyyko** [comment]
See #917 as a previous work of this PR.

**Comment by darcy-shen** [comment]
ci still is failing，you can tune xmake build for macos and mingw

**Comment by darcy-shen** [comment]
i guess that you should remove the `--all` option

**Comment by Oyyko** [comment]
> i guess that you should remove the `--all` option

My fault. Fixed now.

**Comment by Oyyko** [comment]
Actually, I need a way to ask xmake to build every targets except DRAW. I will ask in xmake discussions.

File: src/Mogan/Draw/draw.cpp [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+  immediate_options (argc, argv);
+  load_user_preferences ();
+  string theme= get_user_preference ("gui theme", "default");
+  if (theme == "default") theme= "light";
+  if (theme == "light")
+    tm_style_sheet= "$TEXMACS_PATH/misc/themes/standard-light.css";
+  else if (theme == "dark")
+    tm_style_sheet= "$TEXMACS_PATH/misc/themes/standard-dark.css";
+  else if (theme != "") tm_style_sheet= theme;
+#ifndef OS_MINGW
```
Comment (darcy-shen) [inline src/Mogan/Draw/draw.cpp:530]: 这里也需要移除

File: src/Mogan/Draw/draw.cpp [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+  if (!use_native_menubar) use_unified_toolbar= false;
+  // End user preferences
+
+  if (DEBUG_STD) debug_boot << "Installing internal plug-ins...\n";
+  bench_start ("initialize plugins");
+  init_plugins ();
+  bench_cumul ("initialize plugins");
+  if (DEBUG_STD) debug_boot << "Opening display...\n";
+
+#if defined(X11TEXMACS) && defined(MACOSX_EXTENSIONS)
```
Comment (darcy-shen) [inline src/Mogan/Draw/draw.cpp:356]: 这三行可以直接移除

File: src/Mogan/Draw/draw.cpp [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+#include <locale.h> // for setlocale
+#include <signal.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#ifdef STACK_SIZE
+#include <sys/resource.h>
+#endif
+
+#include "../app_type.hpp"
```
Comment (darcy-shen) [inline src/Mogan/Draw/draw.cpp:23]: 去掉.. 在include里面加上 src/Mogan

**Review by darcy-shen — CHANGES_REQUESTED** [review]

File: xmake_modules/global.lua [hunk 4]
```diff
@@ -0,0 +1,21 @@
+-- The file for Global functions and variables that will be used in the SCRIPT scope of xmake.lua
```
Comment (darcy-shen) [inline xmake_modules/global.lua:1]: 必须是这个目录么 能否放在 misc/xmake这个目录下面

File: src/Mogan/Draw/draw.cpp (hunk 1 above)
Comment (Oyyko) [reply src/Mogan/Draw/draw.cpp:530]: #ifndef means if not defined.

File: xmake_modules/global.lua (hunk 4 above)
Comment (Oyyko) [reply xmake_modules/global.lua:1]: 应该也可以

**Review by darcy-shen — APPROVED** [review]
整体已经ok了 细节需要简单调整一下

**Comment by darcy-shen** [comment]
另外，draw.cpp是新增文件 需要加入到lint里面 用clang-format做格式化

File: .github/workflows/clang-format-linter.yml [hunk 5]
```diff
@@ -20,7 +20,9 @@ on:
       -  'src/Data/Observers/**.cpp'
       -  'src/Data/Observers/**.hpp'
       -  '.github/workflows/clang-format-linter.yml'
-      -  'xmake.lua' 
+      -  'xmake.lua'
+      -  'src/Mogan/**.cpp'
```
Comment (darcy-shen) [inline .github/workflows/clang-format-linter.yml:24]: 这里只是触发ci的文件范围 具体检测哪些文件在下面

File: xmake.lua [hunk 6]
```diff
@@ -95,6 +95,26 @@ else
     add_rules("mode.releasedbg", "mode.release", "mode.debug")
 end
 
+--
+-- global variables in configurations domain
+--
+TeXmacs_files = {
```
Comment (jingkaimori) [inline xmake.lua:101]: `local` keyword could be added for syntax highlighting.

File: xmake.lua [hunk 7]
```diff
@@ -861,4 +916,4 @@ add_configfiles(
             DEVEL_VERSION = DEVEL_VERSION,
         }
     }
-)
+)
```
Comment (jingkaimori) [inline xmake.lua:919]: There should be a return after the end of file

File: misc/xmake/global.lua [hunk 8]
```diff
@@ -0,0 +1,21 @@
+-- The file for Global functions and variables that will be used in the SCRIPT scope of xmake.lua
```
Comment (jingkaimori) [inline misc/xmake/global.lua:1]: Copyright header is needed

**Review by jingkaimori — APPROVED** [review]
LGTM, except only a few format issue needs cleanup

**Pull request merged**